### ISSUE TITLE
Quantization constraint should allow custom ranges

### ISSUE THREAD

**Issue by maurapintor** [pralab/secml-torch#101]
The quantization constraint now works for feature values in [0, 1], but it does not if the values are different.

Planned changes:

```python
class QuantizationConstraint(Constraint):
    """Constraint for ensuring quantized outputs into specified levels."""

    def __init__(self, levels: int, lb: float = 0.0, ub: float = 1.0) -> None:
        """
        Create the QuantizationConstraint.

        Parameters
        ----------
        levels : int
            Number of levels.
        lb: float
            Lower bound for the features (default 0.0).
        ub: float
            Upper bound for the features (default 1.0).
        """
        if int(levels) != levels:
            msg = (
                f"Pass either an integer or a float with no decimals for "
                f"the number of levels (current value: {levels})."
            )
            raise ValueError(msg)

        if lb >= ub:
            msg = f"Lower bound ({lb}) must be less than upper bound ({ub})."
            raise ValueError(msg)

        self.levels = levels
        self.lb = lb
        self.ub = ub
        super().__init__()

    def __call__(self, x: torch.Tensor) -> torch.Tensor:
        """
        Enforce the quantization constraint.

        Parameters
        ----------
        x : torch.Tensor
            Non-quantized input tensor.

        Returns
        -------
        torch.Tensor
            Input with values quantized on the specified
            number of levels.
        """
        # scale x to [0, 1] range
        x_normalized = (x - self.lb) / (self.ub - self.lb)
        # quantize to the number of levels, -1 is to count for the 0
        x_quantized = (x_normalized * (self.levels - 1)).round() / (self.levels - 1)
        # scale back to the original range [lb, ub]
        return x_quantized * (self.ub - self.lb) + self.lb
```

**Comment by maurapintor** [comment]
It's more complicated than that, we should pass the preprocessing items here, as the quantization has to be done directly in the input space. I will come up with a simple way to integrate them in the initialization (I am testing it locally for now).

**Comment by battistabiggio** [comment]
Note that quantization can also be non-uniform. Just saying that you may consider giving a less generic name to the class.

**Comment by maurapintor** [comment]
> Note that quantization can also be non-uniform. Just saying that you may consider giving a less generic name to the class.

Agreed. Thanks! We can create a generic superclass and then make it specific for uniform quantization on an inherited class.

**Comment by maurapintor** [comment]
Opening new related issue (#103) on creating a constraint interface that uses preprocessing. This is useful also for the quantization.

**Comment by maurapintor** [comment]
Implementation plan, with input space constraint, and custom levels:

```python

class QuantizationConstraint(InputSpaceConstraint):
    def __init__(self, preprocessing=None, levels=255) -> None:
        if isinstance(levels, int):
            if levels < 2:
                raise ValueError("Number of levels must be at least 2.")
            # create uniform levels if an integer is provided
            self.levels = torch.linspace(0, 1, levels)
        elif isinstance(levels, list):
            self.levels = torch.tensor(levels, dtype=torch.float32)
        elif isinstance(levels, torch.Tensor):
            self.levels = levels.type(torch.float32)
            if len(self.levels) < 2:
                raise ValueError("Number of custom levels must be at least 2.")
        else:
            raise TypeError("Levels must be an integer, list, or torch.Tensor.")
        # sort levels to ensure they are in ascending order
        self.levels = self.levels.sort().values
        super().__init__(preprocessing)

    def _apply_constraint(self, x):
        # reshape x to facilitate broadcasting with custom levels
        x_expanded = x.unsqueeze(-1)
        # calculate the absolute difference between x and each custom level
        distances = torch.abs(x_expanded - self.levels)
        # find the index of the closest custom level
        nearest_indices = torch.argmin(distances, dim=-1)
        # quantize x to the nearest custom level
        quantized_x = self.levels[nearest_indices]
        return quantized_x

```

**Comment by maurapintor** [comment]
Closed with #104.
